### ISSUE TITLE
siren

### ISSUE THREAD

**Issue by qsobad** [bmild/nerf#60]
will it get better if use the new sirens activation function?

**Comment by tancik** [comment]
We have not yet tested with the sin activation functions proposed in the SIREN paper. Please let us know if you find them beneficial.

**Comment by qsobad** [comment]
if it is as it said in the paper, it is possible to replace positional encoding with relu, so possibly reduce number of input channels here.

**Comment by kwea123** [comment]
@qsobad they recently also researched on other encodings, the results are compelling, you can compare to siren: https://github.com/tancik/fourier-feature-networks

**Comment by qsobad** [comment]
@kwea123 thanks! i read and gaussian seems fine so far
but i still couldn't put sirens to work and reproduce the result it claims. it does not converge all the time and neither does it better than p.e. when converge. i am still investigating what goes wrong.

**Comment by kwea123** [comment]
@qsobad What is your model structure? I also tried siren in NeRF but experienced the same thing as you: its loss is higher and seems to overfit (high val loss)
I tried 7 layers xyz -> sigma and 7 layers xyzdir -> rgb. Maybe I should try the model similar to nerf (use intermediate inputs)

**Comment by qsobad** [comment]
i tried a lot of things, with diff initiation as well. i guess it's nothing about layers, nor skip connection, as from the result of my trials.
my understanding is in siren the weight is the freq, while gaussian specifically pick the weights, so it's harder to control in siren. (it might be wrong.)

**Comment by qsobad** [comment]
i am using
8 layers of sirennet(xyz) + 1 layer & relu -> sigma
8 layers of sirennet(xyz) + 1 intermediate layer + 1 layers of sirennet(dir) & sigmoid -> rgb

**Comment by kwea123** [comment]
Have you tried the Gaussian embedding? I don't find it better either... I modify only the embedding in this repo, so the model structure is a lot more complicated than the experiments done in the paper which is oversimplified (no coarse to fine, no direction embedding...). I tuned lots of parameters, like the standard deviation, the number of embedding sizes, etc, but none of them helps.

It makes me think if the Gaussian only works in the simple settings as in the paper? Like it improves the PSNR from 24 to 25... but in reality 25 is really blurry too. Another possibility is that maybe the embedding is not "friendly" to this NeRF volume rendering? In normal image reconstruction the score is a lot higher than NeRF..

**Comment by qsobad** [comment]
in my case, i use 64 freqencies in gaussian, and i observe clearer edge (not blur), however it has a lot of noise (black dots).

**Comment by kwea123** [comment]
What scene do you use? And how about the psnr score?

**Comment by qsobad** [comment]
i use my own scene, score is worse than pe. mine is:
`
        for freq in self.freq_bands_gaussian:
            for func in self.funcs:
                out += [func(torch.einsum("ij,j->ij", x, freq))]
`

**Comment by kwea123** [comment]
Is your scene forward facing? I'm thinking if it's the NDC part that poses problem. NDC maps the whole space into [-1, 1]^3 but the actual scale is different for x,y and z (especially z), so it would be reasonable to separate the components. Their experiment is on lego which doesn't have this problem.

I guess it's the same, if your `freq` is a vector3 with different components

**Comment by qsobad** [comment]
oh i said it wrong, psnr score is better in gaussian but gaussian is 64 freqs while pe is 10.

my x is vec3 and freq is vec3.

**Comment by kwea123** [comment]
Yes, it's the same as what I find to be better. Have you tried `sin(ax+by+cz)` before? which is their original implementation

**Comment by qsobad** [comment]
i haven't. but i am worried more about the noice. at first, i thought it is due to overfit, however, i did a grid search, even with small number of freq or scale, it is still noisy.

**Comment by kwea123** [comment]
What is your optimal parameters? By the way I also find using a uniform sampling in [0, M] better... since large frequency is indispensable, but Gaussian samples more lower frequencies around 0.
I use uniform sampling in [0, 128] for the frequencies and I find 32 is already sufficient (better than pe), that makes 3*2*32=192 total features though.

**Comment by qsobad** [comment]
64 & 128 is similar, so i picked 64. scale is better in ard 16. i am using uniform sampling in the first place, so i haven't try others.

**Comment by qsobad** [comment]
by uniform sampling, you mean p.e. but different sampling scheme? maybe you can try prime number sampling too?

**Comment by qsobad** [comment]
> What is your optimal parameters? By the way I also find using a uniform sampling in [0, M] better... since large frequency is indispensable (but not too much, it degrades the performance too), but Gaussian samples more lower frequencies around 0.
> I use uniform sampling in [0, 128] for the frequencies and I find 32 samples is already sufficient (better than pe), that makes 3x2x32=192 total features though.

you're right. it turns out that gaussian is very similar to uniform posenc, and better than logscale posenc and prime, when equal number of freq used.

**Comment by qsobad** [comment]
i guess it is due to fourier series:
f(x) = a0/2 + sum_n(a_n * sin(nx)+b_n* cos(nx))

**Comment by tancik** [comment]
It is awesome to see all of the experimentation, there are a number of great observations. 
Our current view is that using the random Fourier features is the best approach in general (no priors on the scene). However if the scene is well aligned to an axis, the positional encoding presented in NeRF will ofter perform better since the frequencies components are all axis aligned. This bias can be observed in scenes like “lego” which is aligned to the axis and preforms better with positional encoding. We also observe that for both methods it is beneficial to increase the number of features for improved accuracy (note that this will also increase the computational cost).

**Comment by kwea123** [comment]
@tancik Thanks for reaching out. How do you tell if a scene is "axis-aligned" or not? For the lego scene it is somewhat easy, but in real scene like "fern", how can you tell if it's axis aligned or not in advance? Or eventually we still need to perform experiments before we can decide which encoding is better?

**Comment by tancik** [comment]
Your intuition is correct that the axis alignment bias is almost exclusively limited to the synthetic scenes. For this reason, the random Fourier features is a good default for real scenes. In the ideal case you would sample the frequencies and directions to match the target scene. We leave these extensions to future work.

**Comment by qsobad** [comment]
i did some more experiments, and i have some observations that doesn't fit the explaination.
i did it on a real scene (one only) so i guess it's not axis aligned.
1. i found that gaussian has much more observable defacts, and actually it has lower score too.
2. i used prime number as well, and it shows details more quickly, but it has lower score that posenc (i guess it's because of missing frequencies).
score: posenc > prime > gaussian (the curve fits perfectly, just around 0.25 higher)

**Comment by kwea123** [comment]
@qsobad Can you share your final siren structure? I tried many different parameters (different number of features, layers, frequencies) but **none of them** gives me any better results... The scores are all 5 points less than PE...

**Comment by qsobad** [comment]
it's very similar to posenc
xyz -> 8*sirennet -> 1*linear + relu -> sigma
xyz -> 8*sirennet -> 1*linear -> 1*sirennet -> 1linear + sigmoid -> rgb

**Comment by sally-chen** [comment]
How do you know if the lego dataset is axis aligned?

**Comment by kwea123** [comment]
@qsobad Disabling some losses make it worse, so I think maybe the reason it doesn't work for NeRF is that the reconstruction l2 loss is not enough to make it learn, we need some other losses.
@sally-chen look at the image to see if the lego car is parallel to the axis or not, this is very subjective and not applicable to real scene.

**Comment by sally-chen** [comment]
@kwea123 Thanks for answering! And just wondering, if there are many frequencies included in positional encoding why is it still only limited to axis aligned scenes? (ie, don't different frequencies represent different axes? Where is the limitation from? )

**Comment by kwea123** [comment]
PE has only features like `sin(ax)`, so the effect of x,y,z is independent. In order to capture non axis aligned features, one would need features like `sin(ax+by+cz)` which mix x,y,z together.

**Comment by qsobad** [comment]
> PE has only features like `sin(ax)`, so the effect of x,y,z is independent. In order to capture non axis aligned features, one would need features like `sin(ax+by+cz)` which mix x,y,z together.

that would be like another direction (sth like [a,b,c]) of fourier feature?

**Comment by sally-chen** [comment]
I see, did not not think from this perspective at all! This is representing random vectors in frequency space vs. on axis vectors. Thanks a lot!

**Comment by qsobad** [comment]
i have a question for the authors, is there any ways to get rid of axis throughly, by describing the geometry with solely the curvature of the field. @tancik

**Comment by qsobad** [comment]
so, the log scale of embedding is coming from wavelet transform?

**Comment by krikru** [comment]
@qsobad Yes, exactly; it would have the direction (a, b, c).

**Comment by krikru** [comment]
@qsobad How did you initialize the weights for the SIREN layers? According to the paper, they propose to draw the weights from the distribution u(-sqrt(6/n), sqrt(6/n)) (where n is the number of neurons in the previous layer), and for the first layer, they use a modified activation function where they multiply Wx by 30 in the argument to the sin function.

**Comment by krikru** [comment]
If you use the `i_embed -1` option to skip input encoding, in what range will the inputs to the network lie?
